unix_integration/src/pam_data.rs

```diff
@@ -0,0 +1,7 @@
+use serde::{Deserialize, Serialize};
+
+/* This is the definition for extra data to be sent along with a pam_prompt
+ * request. It will be sent back to the idprovider to continue an auth attempt.
+ */
+#[derive(Serialize, Deserialize, Debug)]
+pub struct PamData {}
```

Comment:
Rather than a struct, could this be an enum that has say :

```
enum PamData {
    None,
    MFACred( String ),
}
```

Or something similar? Really even with this, I think PamData and PamCred are very "overlapping" here in what they represent being a enum of possible inputs to a pam challenge.

Reply:
In other words, could PamCred::MFACred { value: ..., data }, actually be the enum variant to take the extra data you need in that step only?

Reply:
Except in Himmelblau PamData contains one of these:
```rust
pub struct DeviceAuthorizationFlow {
    pub device_code: String,
    pub user_code: String,
    pub verification_uri: String,
    /* MS doesn't implement verification_uri_complete yet, but our
     * authentication will be simpler once they do, so assume it works and fall
     * back to verification_uri if it doesn't.
     */
    pub verification_uri_complete: Option<String>,
    pub expires_in: u32,
    pub interval: Option<u32>,
    pub message: Option<String>,
}
```

Reply:
What I'm trying to do is ensure this information is kept secret, and only available to that auth attempt. I'm thinking now it makes more sense to keep this in the pam state instead.

Reply:
> In other words, could PamCred::MFACred { value: ..., data }, actually be the enum variant to take the extra data you need in that step only?

It wasn't meant to be extra data received. It's supposed to represent data _maintained_ from one call to step function to the next.

Reply:
Regardless, I do think it makes more sense for this to be an enum, but it should be in the PamState, not in the PamPrompt.

Reply:
It should look something like this, I think:
```rust
enum PamData {
    AwaitingPassword,
    AwaitingMFA(DeviceAuthorizationFlow),
}
```
This way if the idprovider can check if it receives an appropriate response from the pam module (it shouldn't get a password response when trying to do MFA, for example).